File: app/src/main/java/com/productivity/habits/data/local/dao/HabitDao.kt
package com.productivity.habits.data.local.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import androidx.room.Upsert
import com.productivity.habits.data.local.entity.HabitEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface HabitDao {

    @Query("SELECT * FROM habits ORDER BY pinned DESC, createdAt ASC")
    fun getAllHabits(): Flow<List<HabitEntity>>

    @Query("SELECT * FROM habits WHERE archived = 0 ORDER BY pinned DESC, createdAt ASC")
    fun getActiveHabits(): Flow<List<HabitEntity>>

    @Query("SELECT * FROM habits WHERE archived = 0 ORDER BY pinned DESC, createdAt ASC")
    suspend fun getActiveHabitsOnce(): List<HabitEntity>

    @Query("SELECT * FROM habits WHERE archived = 1 ORDER BY updatedAt DESC")
    fun getArchivedHabits(): Flow<List<HabitEntity>>

    @Query("SELECT * FROM habits WHERE pinned = 1 AND archived = 0 ORDER BY createdAt ASC")
    fun getPinnedHabits(): Flow<List<HabitEntity>>

    @Query("SELECT * FROM habits WHERE id = :id")
    fun getHabitById(id: String): Flow<HabitEntity?>

    @Query("SELECT * FROM habits WHERE id = :id")
    suspend fun getHabitByIdOnce(id: String): HabitEntity?

    @Query("SELECT * FROM habits WHERE categoryId = :categoryId AND archived = 0 ORDER BY pinned DESC, createdAt ASC")
    fun getHabitsByCategory(categoryId: String): Flow<List<HabitEntity>>

    @Upsert
    suspend fun upsertHabit(habit: HabitEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertHabits(habits: List<HabitEntity>)

    @Update
    suspend fun updateHabit(habit: HabitEntity)

    @Delete
    suspend fun deleteHabit(habit: HabitEntity)

    @Query("DELETE FROM habits WHERE id = :id")
    suspend fun deleteHabitById(id: String)

    @Query("UPDATE habits SET pinned = :pinned, updatedAt = :updatedAt WHERE id = :id")
    suspend fun updatePinned(id: String, pinned: Boolean, updatedAt: java.time.Instant)

    @Query("UPDATE habits SET archived = :archived, updatedAt = :updatedAt WHERE id = :id")
    suspend fun updateArchived(id: String, archived: Boolean, updatedAt: java.time.Instant)
}

File: lib/data/local/daos/habit_dao.dart
import 'package:drift/drift.dart';
import '../app_database.dart';
import '../tables/habits.dart';

part 'habit_dao.g.dart';

@DriftAccessor(tables: [Habits])
class HabitDao extends DatabaseAccessor<AppDatabase> with _$HabitDaoMixin {
  HabitDao(super.db);

  Stream<List<HabitRow>> watchAllHabits() {
    return (select(habits)
          ..where((h) => h.isDeleted.equals(false))
          ..orderBy([
            (h) => OrderingTerm(expression: h.pinned, mode: OrderingMode.desc),
            (h) => OrderingTerm(expression: h.createdAt, mode: OrderingMode.asc),
          ]))
        .watch();
  }

  Stream<List<HabitRow>> watchActiveHabits() {
    return (select(habits)
          ..where((h) => h.archived.equals(false) & h.isDeleted.equals(false))
          ..orderBy([
            (h) => OrderingTerm(expression: h.pinned, mode: OrderingMode.desc),
            (h) => OrderingTerm(expression: h.createdAt, mode: OrderingMode.asc),
          ]))
        .watch();
  }

  Future<List<HabitRow>> getActiveHabitsOnce() {
    return (select(habits)
          ..where((h) => h.archived.equals(false) & h.isDeleted.equals(false))
          ..orderBy([
            (h) => OrderingTerm(expression: h.pinned, mode: OrderingMode.desc),
            (h) => OrderingTerm(expression: h.createdAt, mode: OrderingMode.asc),
          ]))
        .get();
  }

  Future<List<HabitRow>> getAllHabitsIncludingDeleted() {
    return select(habits).get();
  }

  Stream<List<HabitRow>> watchArchivedHabits() {
    return (select(habits)
          ..where((h) => h.archived.equals(true) & h.isDeleted.equals(false))
          ..orderBy([
            (h) => OrderingTerm(expression: h.updatedAt, mode: OrderingMode.desc),
          ]))
        .watch();
  }

  Stream<List<HabitRow>> watchPinnedHabits() {
    return (select(habits)
          ..where((h) => h.pinned.equals(true) & h.archived.equals(false) & h.isDeleted.equals(false))
          ..orderBy([
            (h) => OrderingTerm(expression: h.createdAt, mode: OrderingMode.asc),
          ]))
        .watch();
  }

  Stream<HabitRow?> watchHabitById(String id) {
    return (select(habits)..where((h) => h.id.equals(id) & h.isDeleted.equals(false)))
        .watchSingleOrNull();
  }

  Future<HabitRow?> getHabitByIdOnce(String id) {
    return (select(habits)..where((h) => h.id.equals(id) & h.isDeleted.equals(false)))
        .getSingleOrNull();
  }

  Stream<List<HabitRow>> watchHabitsByCategory(String categoryId) {
    return (select(habits)
          ..where((h) => h.categoryId.equals(categoryId) & h.archived.equals(false) & h.isDeleted.equals(false))
          ..orderBy([
            (h) => OrderingTerm(expression: h.pinned, mode: OrderingMode.desc),
            (h) => OrderingTerm(expression: h.createdAt, mode: OrderingMode.asc),
          ]))
        .watch();
  }

  Future<void> upsertHabit(HabitsCompanion habit) {
    return into(habits).insertOnConflictUpdate(habit);
  }

  Future<void> insertHabits(List<HabitsCompanion> habitList) async {
    await batch((b) {
      b.insertAllOnConflictUpdate(habits, habitList);
    });
  }

  Future<void> updateHabit(HabitsCompanion habit) {
    return update(habits).replace(habit);
  }

  Future<int> deleteHabitRow(HabitRow habit) {
    return deleteHabitById(habit.id);
  }

  Future<int> deleteHabitById(String id, [DateTime? updatedAt]) {
    final now = (updatedAt ?? DateTime.now()).toUtc();
    return (update(habits)..where((h) => h.id.equals(id))).write(
      HabitsCompanion(
        isDeleted: const Value(true),
        updatedAt: Value(now),
      ),
    );
  }

  Future<int> updatePinned(String id, bool pinned, DateTime updatedAt) {
    return (update(habits)..where((h) => h.id.equals(id))).write(
      HabitsCompanion(
        pinned: Value(pinned),
        updatedAt: Value(updatedAt),
      ),
    );
  }

  Future<int> updateArchived(String id, bool archived, DateTime updatedAt) {
    return (update(habits)..where((h) => h.id.equals(id))).write(
      HabitsCompanion(
        archived: Value(archived),
        updatedAt: Value(updatedAt),
      ),
    );
  }
}
